api/pkg/controller/ipam/ipam_controller.go

```
@@ -0,0 +1,243 @@
+		glog.V(4).Infof("Machine %s doesn't need initialization", machine.Name)
+		return nil
+	}
+
+	cfg, err := providerconfig.GetConfig(machine.Spec.ProviderSpec)
+	if err != nil {
+		return err
+	}
+
+	ip, network, err := r.getNextFreeIP(ctx)
+	if err != nil {
+		return err
+	}
+
+	mask, _ := network.IPNet.Mask.Size()
+	cidr := fmt.Sprintf("%s/%d", ip.String(), mask)
+
+	cfg.Network = &providerconfig.NetworkConfig{
+		CIDR:    cidr,
+		Gateway: network.Gateway.String(),
+		DNS: providerconfig.DNSConfig{
+			Servers: r.ipsToStrs(network.DNSServers),
+		},
+	}
+
+	cfgSerialized, err := json.Marshal(cfg)
+	if err != nil {
+		return err
+	}
+
+	machine.Spec.ProviderSpec.Value = &runtime.RawExtension{Raw: cfgSerialized}
+	newAnnotationVal := strings.Replace(machine.Annotations[annotationMachineUninitialized],
+		annotationValue,
+		"", -1)
+	machine.Annotations[annotationMachineUninitialized] = newAnnotationVal
+	if err := r.Update(ctx, machine); err != nil {
+		return fmt.Errorf("failed to update machine %q after adding network: %v", machine.Name, err)
+	}
+
+	// Block until the change is in the lister to make sure we don't hand out an IP twice
+	return wait.Poll(1*time.Second, 60*time.Second, func() (bool, error) {
+		newMachine := &clusterv1alpha1.Machine{}
+		if err := r.Get(ctx, types.NamespacedName{Namespace: machine.Namespace, Name: machine.Name}, newMachine); err != nil {
+			return false, err
+		}
+		newConfig, err := providerconfig.GetConfig(newMachine.Spec.ProviderSpec)
+		if err != nil {
+			return false, nil
+		}
+		return newConfig.Network != nil && newConfig.Network.CIDR == cidr, nil
+	})
+}
+
+func (r *reconciler) getNextFreeIP(ctx context.Context) (net.IP, Network, error) {
+	usedIps, err := r.getUsedIPs(ctx)
+	if err != nil {
+		return nil, Network{}, err
+	}
+
+	for _, cidr := range r.cidrRanges {
+		ip, err := r.getNextFreeIPForCIDR(cidr, usedIps)
+		if err == nil {
+			return ip, cidr, nil
+		}
+	}
+
+	return nil, Network{}, cidrExhaustedError{}
```

why not simply return a standard error (`errors.New()`)? From what I have seen there is no code that makes use of `cidrExhaustedError` type.